### PR TITLE
Add `sudo::remove_key`

### PR DESCRIPTION
**Pull request by ggwpez**
Changes:
- Adds a new call `remove_key` to the sudo pallet to permanently remove the sudo key.
- Remove some clones and general maintenance

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4218686

**Review by liamaharon — APPROVED** [review]

**Comment by jasl** [comment]
What about adding a `remove_key` to remove `root`?
By default, `polkadot.js` sets `None` for optional fields. I guess a tiny probability set the `sudo key` to `None` accidentally and unexpectedly, and the result is critical.
A `remove_key` is more explicit.

**Comment by ggwpez** [comment]
> What about adding a `remove_key` to remove `root`? By default, `polkadot.js` sets `None` for optional fields. I guess a tiny probability set the `sudo key` to `None` accidentally and unexpectedly, and the result is critical. A `remove_key` is more explicit.

Yea I also thought about this. But am not sure if we should write our Rust code in such a fashion only to appease a single downstream UI with insane default values...

File: substrate/frame/sudo/src/lib.rs [hunk 1]
```diff
@@ -345,8 +325,19 @@ pub mod pallet {
...
+	impl<T: Config> Pallet<T> {
+		/// Ensure that the caller is the sudo key.
+		pub(crate) fn ensure_sudo(origin: OriginFor<T>) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			if Self::key() != Some(sender) {
+				Err(Error::<T>::RequireSudo.into())
+			} else {
+				Ok(())
 			}
```
Comment (bkchr) [inline substrate/frame/sudo/src/lib.rs:341]:
```suggestion
			if Self::key().map_or(false, |k| k == sender) {
				Ok(())
			} else {
				Err(Error::<T>::RequireSudo.into())
			}
```

File: substrate/frame/sudo/src/lib.rs [hunk 2]
```diff
@@ -222,47 +216,37 @@ pub mod pallet {
...
-		/// The dispatch origin for this call must be _Signed_.
-		///
-		/// ## Complexity
-		/// - O(1).
 		#[pallet::call_index(2)]
 		#[pallet::weight(T::WeightInfo::set_key())]
 		pub fn set_key(
 			origin: OriginFor<T>,
-			new: AccountIdLookupOf<T>,
+			new: Option<AccountIdLookupOf<T>>,
```
Comment (bkchr) [inline substrate/frame/sudo/src/lib.rs:242]: Yeah, I would go with what @jasl said and introduce a `remove_key` function

**Review by bkchr — CHANGES_REQUESTED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4228297

**Review by bkchr — APPROVED** [review]
Nice!

**Review by lexnv — APPROVED** [review]

**Comment by ggwpez** [comment]
bot bench -v PIPELINE_SCRIPTS_REF=oty-fix-runtimes substrate-pallet --pallet=pallet_sudo  
bot bench -v PIPELINE_SCRIPTS_REF=oty-fix-runtimes polkadot-pallet --pallet=pallet_sudo --runtime=westend  
bot bench -v PIPELINE_SCRIPTS_REF=oty-fix-runtimes polkadot-pallet --pallet=pallet_sudo --runtime=rococo

**Pull request merged**